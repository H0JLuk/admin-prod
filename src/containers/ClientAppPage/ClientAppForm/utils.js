import { notification } from 'antd';
import { addSettings, updateSettingsList } from '../../../api/services/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';

export const showNotify = (message, isError) => {
    const config = {
        duration: 5,
        placement: 'bottomRight',
    };

    if (isError) {
        notification.error({
            ...config,
            message,
        });
    } else {
        notification.success({
            ...config,
            message,
        });
    }

};

export async function createOrUpdateKey (changedParams) {
    const updateSettings = [];
    const addSettingsArr = [];

    changedParams.forEach(({ type, ...params }) => {
        type === SETTINGS_TYPES.EDIT
            ? updateSettings.push(params)
            : addSettingsArr.push(params);
    });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export const checkExistDesignSettings = (settings) => {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
};

export const getImage = (imageName) => {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
};


