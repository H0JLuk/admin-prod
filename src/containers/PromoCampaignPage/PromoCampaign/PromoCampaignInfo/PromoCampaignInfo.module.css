.wrapper {
    width: 100%;
    height: 100%;
}

.header {
    position: sticky;
    top: 63px;
    z-index: 1;
    background-color: #ffffff;
    padding: 15px;
    border-bottom: 1px solid #e5e8eb;
}

.promoContainer {
    display: flex;
}

.title {
    font-weight: bold;
    font-size: 20px;
    line-height: 28px;
    letter-spacing: -.38px;
    margin-bottom: 15px;
}

.infoTitle {
    margin-bottom: 7px;
    font-size: 16px;
    font-weight: bold;
    line-height: 26px;
    letter-spacing: -0.3px;
}

.infoText {
    font-size: 14px;
    color: #4e4e4e;
}

.infoDetail {
    padding-bottom: 15px;
}

.containerStep {
    padding: 15px;
}

.container {
    background-color: #f8f8f8;
    padding: 20px 30px;
    height: 100%;
}

.container + .infoDetail {
    margin-top: 15px;
}

.flex {
    display: flex;
    flex-direction: column;
    width: 80%;
}

.flexRow {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    width: 80%;
    max-width: 250px;
}

.checkbox:not(:last-child) {
    margin-bottom: 10px;
}

.switchContainer {
    max-width: 280px;
    display: flex;
    align-items: flex-start;
}

.switch {
    display: flex;
    align-items: center;
}

.switch > button {
    align-self: center;
}

.status {
    display: inline-block;
    padding: 5px 16px;
    background-color: rgba(9, 165, 82, 0.08);
    color: #09a552;
    border-radius: 25px;
    text-align: center;
}

.danger {
    background-color: #fff1f0;
    color: #dc283a;
}

.webLink {
    white-space: nowrap;
    text-overflow: ellipsis;
    max-width: 500px;
    overflow: hidden;
}

.buttonGroup {
    margin-top: 15px;
}

.buttonGroup button:not(:last-child) {
    margin-right: 15px;
}
