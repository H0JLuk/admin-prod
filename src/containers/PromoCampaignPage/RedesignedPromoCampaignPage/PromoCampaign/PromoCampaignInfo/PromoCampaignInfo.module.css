.wrapper {
    width: 100%;
    height: 100%;
}

.header {
    padding: 22px 15px 16px;
    border-bottom: 1px solid #e5e8eb;
}

.promoContainer {
    display: flex;
}

.title {
    font-weight: bold;
    font-size: 20px;
    line-height: 28px;
    letter-spacing: -.38px;
    margin-bottom: 15px;
}

.infoTitle {
    margin-bottom: 7px;
    font-size: 16px;
    font-weight: bold;
    line-height: 26px;
    letter-spacing: -0.3px;
}

.infoText {
    font-size: 14px;
    color: #4e4e4e;
}

.infoDetail {
    margin: 10px 0;
}

.containerStep {
    padding: 0 15px;
}

.container {
    background-color: #f8f8f8;
    padding: 20px 30px;
    height: 100%;
}

.flex {
    display: flex;
    flex-direction: column;
    width: 80%;
}

.checkbox {
    margin: 10px 0;
}

.status {
    display: inline-block;
    padding: 5px 16px;
    background-color: rgba(9, 165, 82, 0.08);
    color: #09a552;
    border-radius: 25px;
    text-align: center;
}

.redStatus {
    background-color:  #fff1f0 !important;
    color: #dc283a !important;
}

.webLink {
    white-space: nowrap;
    text-overflow: ellipsis;
    max-width: 500px;
    overflow: hidden;
}

.buttonGroup {
    margin-top: 15px;
}

.buttonGroup button:not(:last-child) {
    margin-right: 15px;
}

.promoContainer :global(.ant-btn-primary) {
    background-color: #09a552;
    border-color: #09a552;
}

.deleteBtn:global(.ant-btn-primary) {
    background-color: #dd283b;
    border-color: #dd283b;
}