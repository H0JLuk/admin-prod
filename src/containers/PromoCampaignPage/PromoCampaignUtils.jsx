import React from 'react';
import { deletePromoCampaign } from '../../api/services/promoCampaignService';
import { customNotifications } from '../../utils/notifications';


export const getDeleteTitleConfirm = (promoCampaignName) => (
    <span>
        Вы уверены, что хотите удалить промо-кампанию <b>{ promoCampaignName }</b>?
    </span>
);

export const getSuccessDeleteMessage = (promoCampaignName) => (
    <span>
        Промо-кампания <b>{ promoCampaignName }</b> успешно удалена
    </span>
);

export async function onConfirmDeletePromoCampaign(id, name) {
    try {
        await deletePromoCampaign(id);
        customNotifications.success({
            duration: 10,
            message: getSuccessDeleteMessage(name),
        });
    } catch ({ message }) {
        customNotifications.error({
            message,
        });
        throw new Error(message);
    }
}
