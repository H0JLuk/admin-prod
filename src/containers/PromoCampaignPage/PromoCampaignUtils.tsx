import React from 'react';
import { deletePromoCampaign } from '../../api/services/promoCampaignService';
import { customNotifications } from '../../utils/notifications';

export const getDeleteTitleConfirm = (promoCampaignName: string) => (
    <span>
        Вы уверены, что хотите удалить промо-кампанию <b>{promoCampaignName}</b>?
    </span>
);

export const getSuccessDeleteMessage = (promoCampaignName: string) => (
    <span>
        Промо-кампания <b>{promoCampaignName}</b> успешно удалена
    </span>
);

export async function onConfirmDeletePromoCampaign(id: number, name: string) {
    try {
        await deletePromoCampaign(id);
        customNotifications.success({
            duration: 10,
            message: getSuccessDeleteMessage(name),
        });
    } catch ({ message }) {
        customNotifications.error({
            message,
        });
        throw new Error(message);
    }
}
