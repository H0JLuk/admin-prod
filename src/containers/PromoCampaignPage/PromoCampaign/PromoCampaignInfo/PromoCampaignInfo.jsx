import React, { useEffect, useState, useCallback, useMemo } from 'react';
import { generatePath, useHistory, useLocation } from 'react-router-dom';
import { Button } from 'antd';
import { getFormattedDate } from '../../../../utils/helper';
import { PROMO_CAMPAIGN_PAGES } from '../../../../constants/route';
import Header from '../../../../components/Header/Header';
import PromoCampaignSideBar from '../PromoCampaignSideBar/PromoCampaignSideBar';
import StepInfo from './Steps/StepInfo';
import StepTextAndImage from './Steps/StepTextAndImage';
import StepVisibility from './Steps/StepVisibility';
import PromoCampaignCopyModal from '../PromoCampaignCopyModal';
import { confirmModal } from '../../../../utils/utils';
import { onConfirmDeletePromoCampaign, getDeleteTitleConfirm } from '../../PromoCampaignUtils';

import styles from './PromoCampaignInfo.module.css';

const EDIT = 'Редактировать';
const COPY = 'Копировать';
const DELETE = 'Удалить';

const PromoCampaignInfo = ({ matchUrl }) => {
    const history = useHistory();
    const [promoCampaign, setPromoCampaign] = useState({});
    const { state: stateFromLocation } = useLocation();
    const [step, setStep] = useState(0);

    useEffect(() => {
        const { promoCampaign } = stateFromLocation || {};
        if (!promoCampaign) {
            return history.push(matchUrl);
        }
        setPromoCampaign({
            ...promoCampaign,
            startDate: getFormattedDate(promoCampaign.startDate),
            finishDate: getFormattedDate(promoCampaign.finishDate),
        });
        setStep(1);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onEdit = useCallback(() => {
        const path = `${ matchUrl }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`;
        const { promoCampaign } = stateFromLocation || {};
        history.push(generatePath(path, { promoCampaignId: promoCampaign.id }), { promoCampaign });
    }, [history, matchUrl, stateFromLocation]);

    const onConfirmDelete = useCallback(async () => {
        try {
            await onConfirmDeletePromoCampaign(promoCampaign.id, promoCampaign.name);
            history.push(matchUrl);
        } catch (e) {
            console.warn(e);
        }
    }, [promoCampaign.id, promoCampaign.name, history, matchUrl]);

    const onDeleteClick = useCallback(() => confirmModal({
        onOk: onConfirmDelete,
        title: getDeleteTitleConfirm(promoCampaign.name),
        centered: false,
    }), [promoCampaign.name, onConfirmDelete]);

    const StepContainer = useMemo(() => {
        switch (step) {
            case 1:
                return <StepInfo promoCampaign={ promoCampaign } />;
            case 2:
                return (
                    <StepTextAndImage
                        type={ promoCampaign.type }
                        banners={ promoCampaign.banners }
                        texts={ promoCampaign.texts }
                    />
                );
            case 3:
                return <StepVisibility />;
            default:
                return null;
        }
    }, [step, promoCampaign]);

    return (
        <div className={ styles.promoContainer }>
            <PromoCampaignSideBar
                active={ step }
                onClick={ setStep }
            />
            <div className={ styles.wrapper }>
                <Header />
                { step !== 3 && (
                    <div className={ styles.header }>
                        <div className={ styles.title }>{ promoCampaign?.name }</div>
                        <div className={ styles.buttonGroup }>
                            <Button
                                type="primary"
                                onClick={ onEdit }
                            >
                                { EDIT }
                            </Button>
                            <PromoCampaignCopyModal promoCampaignData={ stateFromLocation?.promoCampaign }>
                                <Button type="primary">
                                    { COPY }
                                </Button>
                            </PromoCampaignCopyModal>
                            <Button
                                type="primary"
                                danger
                                onClick={ onDeleteClick }
                            >
                                { DELETE }
                            </Button>
                        </div>
                    </div>
                ) }
                { StepContainer }
            </div>
        </div>
    );
};

export default PromoCampaignInfo;
