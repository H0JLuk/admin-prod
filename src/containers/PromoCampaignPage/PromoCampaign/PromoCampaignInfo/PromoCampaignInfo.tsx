import React, { useEffect, useState, useCallback, useMemo } from 'react';
import { generatePath, useHistory, useLocation } from 'react-router-dom';
import { Button } from 'antd';
import { getFormattedDate } from '@utils/helper';
import { PROMO_CAMPAIGN_PAGES } from '@constants/route';
import Header from '@components/Header';
import PromoCampaignSideBar from '../PromoCampaignSideBar';
import StepInfo from './Steps/StepInfo';
import StepTextAndImage from './Steps/StepTextAndImage';
import StepVisibility from './Steps/StepVisibility';
import PromoCampaignCopyModal from '../PromoCampaignCopyModal';
import { confirmModal } from '@utils/utils';
import { onConfirmDeletePromoCampaign, getDeleteTitleConfirm } from '../../PromoCampaignUtils';
import { BUTTON_TEXT } from '@constants/common';
import { PromoCampaignDto } from '@types';

import styles from './PromoCampaignInfo.module.css';

type PromoCampaignInfoProps = {
    matchPath: string;
};

interface ILocationState {
    promoCampaign: PromoCampaignDto;
}

const PromoCampaignInfo: React.FC<PromoCampaignInfoProps> = ({ matchPath }) => {
    const history = useHistory();
    const [promoCampaign, setPromoCampaign] = useState({} as PromoCampaignDto);
    const { state: stateFromLocation } = useLocation<ILocationState>();
    const [step, setStep] = useState(0);

    useEffect(() => {
        const { promoCampaign: promoCampaignFromLocation } = stateFromLocation || {};
        if (!promoCampaignFromLocation) {
            return history.push(matchPath);
        }
        setPromoCampaign({
            ...promoCampaignFromLocation,
            startDate: getFormattedDate(promoCampaignFromLocation.startDate),
            finishDate: getFormattedDate(promoCampaignFromLocation.finishDate),
        });
        setStep(1);
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    const onEdit = useCallback(() => {
        const path = `${ matchPath }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`;
        const { promoCampaign: promoCampaignFromLocation } = stateFromLocation || {};
        history.push(generatePath(path, { promoCampaignId: promoCampaignFromLocation.id }), { promoCampaign: promoCampaignFromLocation });
    }, [history, matchPath, stateFromLocation]);

    const onConfirmDelete = useCallback(async () => {
        try {
            await onConfirmDeletePromoCampaign(promoCampaign.id, promoCampaign.name);
            history.push(matchPath);
        } catch (e) {
            console.warn(e);
        }
    }, [promoCampaign.id, promoCampaign.name, history, matchPath]);

    const onDeleteClick = useCallback(() => confirmModal({
        onOk: onConfirmDelete,
        title: getDeleteTitleConfirm(promoCampaign.name),
        centered: false,
    }), [promoCampaign.name, onConfirmDelete]);

    const StepContainer = useMemo(() => {
        switch (step) {
            case 1:
                return <StepInfo promoCampaign={promoCampaign} />;
            case 2:
                return (
                    <StepTextAndImage
                        type={promoCampaign.type}
                        banners={promoCampaign.banners}
                        texts={promoCampaign.texts}
                    />
                );
            case 3:
                return <StepVisibility />;
            default:
                return null;
        }
    }, [step, promoCampaign]);

    return (
        <div className={styles.promoContainer}>
            <PromoCampaignSideBar
                active={step}
                onClick={setStep}
            />
            <div className={styles.wrapper}>
                <Header />
                {step !== 3 && (
                    <div className={styles.header}>
                        <div className={styles.title}>{promoCampaign?.name}</div>
                        <div className={styles.buttonGroup}>
                            <Button
                                type="primary"
                                onClick={onEdit}
                            >
                                {BUTTON_TEXT.EDIT}
                            </Button>
                            <PromoCampaignCopyModal promoCampaignData={stateFromLocation?.promoCampaign}>
                                <Button type="primary">
                                    {BUTTON_TEXT.COPY}
                                </Button>
                            </PromoCampaignCopyModal>
                            <Button
                                type="primary"
                                danger
                                onClick={onDeleteClick}
                            >
                                {BUTTON_TEXT.DELETE}
                            </Button>
                        </div>
                    </div>
                )}
                {StepContainer}
            </div>
        </div>
    );
};

export default PromoCampaignInfo;
