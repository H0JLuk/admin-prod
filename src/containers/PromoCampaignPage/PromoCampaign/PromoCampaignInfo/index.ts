import PromoCampaignInfo from './PromoCampaignInfo';

export default PromoCampaignInfo;
