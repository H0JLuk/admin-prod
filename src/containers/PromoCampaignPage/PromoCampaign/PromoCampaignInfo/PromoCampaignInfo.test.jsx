import React from 'react';
import { generatePath } from 'react-router-dom';
import { shallow } from 'enzyme';
import { render, fireEvent, screen } from '@testing-library/react';
import { useLocation } from 'react-router';
import PromoCampaignInfo from './PromoCampaignInfo';
import { confirmModal } from '../../../../utils/utils';
import { getDeleteTitleConfirm, onConfirmDeletePromoCampaign } from '../../PromoCampaignUtils';
import * as clientAppService from '../../../../api/services/clientAppService';
import { clientAppTestData, promoCampaignTestData } from '../../../../../__tests__/constants';
import { PROMO_CAMPAIGN_PAGES } from '../../../../constants/route';
import { sleep } from '../../../../setupTests';

const state = {
    promoCampaign: promoCampaignTestData,
};

jest.mock('../../PromoCampaignUtils', () => ({
    getDeleteTitleConfirm: jest.fn(),
    onConfirmDeletePromoCampaign: jest.fn(),
}));

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useHistory: () => ({
        push: mockHistoryPush,
    }),
    generatePath: jest.fn(),
}));

jest.mock('react-router', () => ({
    useLocation: jest.fn(),
}));

jest.mock('../../../../utils/utils', () => ({
    confirmModal: jest.fn(),
}));

beforeEach(() => {
    clientAppService.getClientAppInfo = jest.fn().mockResolvedValue(clientAppTestData);
    useLocation.mockImplementation(() => ({
        state,
    }));
});

describe('<PromoCampaignInfo /> tests', () => {
    const props = {
        matchPath: '/admin/promo-campaign',
        history: {
            push: jest.fn(),
        },
    };

    it('PromoCampaignInfo snapshot', () => {
        const PromoCampaignInfoItem = shallow(<PromoCampaignInfo { ...props } />);
        expect(PromoCampaignInfoItem.debug()).toMatchSnapshot();
    });

    it('should render stepInfo', () => {
        const wrapper = shallow(<PromoCampaignInfo { ...props } />);
        wrapper.find('PromoCampaignSideBar').simulate('click', 1);
        expect(wrapper.find('StepInfo')).toHaveLength(1);
    });

    it('should render TextAndImageStep', () => {
        const wrapper = shallow(<PromoCampaignInfo { ...props } />);
        wrapper.find('PromoCampaignSideBar').simulate('click', 2);
        expect(wrapper.find('StepTextAndImage')).toHaveLength(1);
    });

    it('should render StepVisibility', () => {
        const wrapper = shallow(<PromoCampaignInfo { ...props } />);
        wrapper.find('PromoCampaignSideBar').simulate('click', 3);
        expect(wrapper.find('StepVisibility')).toHaveLength(1);
    });

    it('should call confirmModal when delete btn click', () => {
        const wrapper = shallow(<PromoCampaignInfo { ...props } />);
        wrapper.find('PromoCampaignSideBar').simulate('click', 1);
        wrapper.find('Button').last().simulate('click');
        expect(confirmModal).toBeCalled();
        expect(getDeleteTitleConfirm).toBeCalled();
    });

    it('should render PromoCampaignInfo with promoCampaign data', async () => {
        render(<PromoCampaignInfo { ...props } />);
        await sleep();
        const title = screen.getByText(state.promoCampaign.name);
        expect(title.textContent).toEqual(state.promoCampaign.name);
    });

    it('should redirect to promo campaign page', async () => {
        console.error = jest.fn();
        useLocation.mockImplementation(() => ({
            state: {
                promoCampaign: undefined,
            },
        }));
        render(<PromoCampaignInfo { ...props } />);
        await sleep();
        expect(mockHistoryPush).toBeCalledWith(props.matchPath);
    });

    it('should fire onDeleteClick/confirmModal fn by click', async () => {
        const { getByText } = render(<PromoCampaignInfo { ...props } />);
        await sleep();
        fireEvent.click(getByText('Удалить'));
        expect(confirmModal).toBeCalled();
        await confirmModal.mock.calls[0][0].onOk();
        expect(onConfirmDeletePromoCampaign).toBeCalledWith(state.promoCampaign.id, state.promoCampaign.name);
        expect(mockHistoryPush).toBeCalledWith(props.matchPath);
    });

    it('should warn with error if onConfirmDeletePromoCampaign rejected', async () => {
        onConfirmDeletePromoCampaign.mockRejectedValueOnce('error');
        const spy = jest.spyOn(console, 'warn').mockImplementation(() => '');
        const { getByText } = render(<PromoCampaignInfo { ...props } />);
        await sleep();
        fireEvent.click(getByText('Удалить'));
        await confirmModal.mock.calls[0][0].onOk();
        expect(spy).toHaveBeenCalled();
    });

    it('should fire onEdit callback by click', async () => {
        generatePath.mockImplementationOnce((path, { promoCampaignId }) => `${path}/${promoCampaignId}`);
        const path = `${ props.matchPath }${ PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT }`;
        const { getByText } = render(<PromoCampaignInfo { ...props } />);
        await sleep();
        fireEvent.click(getByText('Редактировать'));

        expect(mockHistoryPush).toBeCalledWith(`${props.matchPath}${PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT}/${state.promoCampaign.id}`, { promoCampaign: state.promoCampaign });
        expect(generatePath).toBeCalledWith(path, { promoCampaignId: state.promoCampaign.id });
        expect(mockHistoryPush).toBeCalled();
    });
});
