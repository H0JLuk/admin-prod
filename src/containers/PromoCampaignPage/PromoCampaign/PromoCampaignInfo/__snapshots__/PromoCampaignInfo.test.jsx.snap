// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromoCampaignInfo /> tests PromoCampaignInfo snapshot 1`] = `
"<div className=\\"promoContainer\\">
  <PromoCampaignSideBar active={0} onClick={[Function: bound _dispatchAction]} />
  <div className=\\"wrapper\\">
    <Header />
    <div className=\\"header\\">
      <div className=\\"title\\" />
      <div className=\\"buttonGroup\\">
        <Button type=\\"primary\\" onClick={[Function (anonymous)]}>
          Редактировать
        </Button>
        <PromoCampaignCopyModal promoCampaignData={{...}}>
          <Button type=\\"primary\\">
            Копировать
          </Button>
        </PromoCampaignCopyModal>
        <Button type=\\"primary\\" danger={true} onClick={[Function (anonymous)]}>
          Удалить
        </Button>
      </div>
    </div>
  </div>
</div>"
`;
